service/src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java

```diff
@@ -115,6 +123,36 @@ private String getResourceLocation(Workspace workspace, String requestedLocation
         : requestedLocation;
   }
 
+  private @Nullable String getResourceZone(Workspace workspace, String requestedZone) {
```

Comment:
This computation should be done as part of the createNotebook flight, not in the controller. We try to avoid cloud calls and external service calls from the controller. Plus by doing them in a Stairway flight, we get to easily retry on cloud errors.

I think you should use `getResourceLocation` above to fill in the location param in the object.
Then do the part from line 135 on down in the `RetrieveNetworkNameStep`. That isolates the zone concerns to that one place.

Reply:
Hey @ddietterich , sorry for the late reply. The issue I'm running into when trying this is that there doesn't seem to be a good way to store back the computed zone in the step. Since the API allows you to specify a zone, using getResourceLocation would mean mixing zones/regions in the location param and while we could store the real zone in the flight's working map, the other operations like delete or update don't have access to this info.

Reply:
I see, you need to populate the resource object with the corrected location. And that you need to do in the controller.

Reply:
You can store back the computed zone by calling the resourceDao with updated attributes.
I would suggest the following structure:
- In `RetrieveNetworkNameStep` remember the complete region including zone in the working map.
- Add a step `UpdateMetadataRegionStep`. The constructor would take the AI Notebook resource object (this is an input parameter) and the `ResourceDao` bean. The do() code would call `updateResource` providing a revised set of attributes that include the region.

This is not a perfect solution, but it is closer to the way I am hoping to move the other resources. So it will become more perfect over time :)

Reply:
To expand a bit. We have a general design problem with handling the per-resource attributes. It is done inconsistently. Having to have completely built resources in the controller in the create case makes things like this messy. It also makes clone messy, since we are building the resources from several different pieces.

So sorry you plunged into one of the messier parts of WSM...

Reply:
TYSM for the pointers.